src/Type/Php/ArrayFirstDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,48 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\DependencyInjection\AutowiredService;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\NullType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+
+#[AutowiredService]
+final class ArrayFirstDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_first' && $functionReflection->isBuiltin();
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		$args = $functionCall->getArgs();
+
+		if (count($args) < 1) {
+			return null;
+		}
+
+		$argType = $scope->getType($args[0]->value);
+		$iterableAtLeastOnce = $argType->isIterableAtLeastOnce();
+
+		if ($iterableAtLeastOnce->no()) {
+			return new NullType();
+		}
+
+		$valueType = $argType->getFirstIterableValueType();
```

Comment:
When I thought about the implementation, I'd do only `getIterableValueType`. The `getFirstIterable*Type` methods are highly misleading and we should get rid of it. Because array shapes do not enforce the order of the keys. Sorry to put more burden on you, but can you please deprecate `Type::getFirstIterableKeyType`, `getLastIterableKeyType`, `getFirstIterableValueType`, `getLastIterableValueType` and use `getIterableKeyType` and `getIterableValueType` in all places instead? And just update any test assertions because of that. Thank you!

Reply:
Alright. I can do that. But we will lose the precise type for constant arrays that way.